src/Scope/TreeTraverser.php

```diff
@@ -0,0 +1,193 @@
+            }
+
+            if ($node instanceof Node\Expression\AnonymousFunctionCreationExpression
+                && $node->anonymousFunctionUseClause !== null
+                && $node->anonymousFunctionUseClause->useVariableNameList !== null) {
+                foreach ($node->anonymousFunctionUseClause->useVariableNameList->getElements() as $use) {
+                    $name = $use->getName();
+                    // Used variable in an anonymous function. Same as parent type, Mixed if not defined in parent.
+                    $childScope->variables[$name] = new Variable(
+                        isset($scope->variables[$name]) ? $scope->variables[$name]->type : new Types\Mixed_,
+                        $use
+                    );
+                }
+            }
+
+            return $childScope;
+        }
+
+        if ($node instanceof ClassLike
+            && (in_array($childName, ['classMembers', 'interfaceMembers','traitMembers'], true))
+        ) {
+            $childScope = new Scope;
+            $childScope->resolvedNameCache = $scope->resolvedNameCache;
+            $thisVar = new Variable(
+                new Types\Object_(new Fqsen('\\' . (string)$node->getNamespacedName())),
+                $node
+            );
+            $childScope->variables['this'] = $thisVar;
+            $childScope->currentSelf = $thisVar;
+            return $childScope;
+        }
+
+        return $scope;
+    }
+
+    /**
+     * Adds any variables declared by $node to $scope.
+     *
+     * Note that functions like extract and parse_str are not handled.
+     *
+     * @return void
+     */
+    private function modifyScopeAfterNode(Node $node, Scope $scope)
+    {
+        if ($node instanceof Expression\AssignmentExpression) {
+            if ($node->operator->kind !== TokenKind::EqualsToken
+                || !$node->leftOperand instanceof Expression\Variable
+                || $node->rightOperand === null
+                || $node->rightOperand instanceof MissingToken
+            ) {
+                return;
+            }
+            $scope->variables[$node->leftOperand->getName()] = new Variable(
+                $this->definitionResolver->resolveExpressionNodeToType($node->rightOperand, $scope),
+                $node
+            );
+        } else if (($node instanceof Node\ForeachValue || $node instanceof Node\ForeachKey)
+            && $node->expression instanceof Node\Expression\Variable
+        ) {
+            $scope->variables[$node->expression->getName()] = new Variable(
+                $this->definitionResolver->getTypeFromNode($node, $scope),
+                $node
+            );
+        } else if ($node instanceof Node\Statement\NamespaceDefinition) {
+            // After a new namespace A\B;, the current alias table is flushed.
+            $scope->clearResolvedNameCache();
+        }
+
+
+        // TODO: Handle use (&$x) when $x is not defined in scope.
+        // TODO: Handle list(...) = $a;
```

Comment:
and PHP7+ `[...] = $a;`

Reply:
:+1: I didn't even know about that construct

Reply:
Ok, I added a TODO comment for that and `foreach ($a as [...])`